src/binary/timestamp.rs

```diff
@@ -11,75 +11,89 @@ use crate::{
     types::timestamp::{Precision, Timestamp},
 };
 
-const ENCODED_BUFFER_SIZE: usize = 16;
+const MAX_TIMESTAMP_LENGTH: usize = 16;
 
 /// Extends [`Timestamp`] to support [Ion binary].
 ///
 /// [Ion binary]: https://amzn.github.io/ion-docs/docs/binary.html#6-timestamp
-pub trait TimestampBinaryExt {
-    fn encode_binary(&self) -> IonResult<ArrayVec<u8, { ENCODED_BUFFER_SIZE }>>;
-    fn write_binary<W: Write>(&self, sink: &mut W) -> IonResult<()>;
+// TODO: Change these methods to return the number of bytes written.
+pub trait TimestampBinaryEncoder {
+    /// Encodes the content of a [`Timestamp`] as per the Ion binary encoding.
+    /// Returns the length of the encoded bytes.
+    ///
+    /// This does not encode the type descriptor nor the associated length.
+    /// Prefer [`TimestampBinaryEncoder::encode_timestamp_value`] for that.
+    fn encode_timestamp(&mut self, timestamp: &Timestamp) -> IonResult<()>;
+
+    /// Encodes a [`Timestamp`] as an Ion value with the type descriptor and length.
+    /// Returns the length of the encoded bytes.
+    fn encode_timestamp_value(&mut self, timestamp: &Timestamp) -> IonResult<()>;
 }
 
-impl TimestampBinaryExt for Timestamp {
+impl<W> TimestampBinaryEncoder for W
+where
+    W: Write,
+{
     /// NOTE: Currently, this function always encodes with nanosecond precision.
-    fn encode_binary(&self) -> IonResult<ArrayVec<u8, { ENCODED_BUFFER_SIZE }>> {
+    fn encode_timestamp(&mut self, timestamp: &Timestamp) -> IonResult<()> {
         const SECONDS_PER_MINUTE: f32 = 60f32;
-        let mut sink = ArrayVec::new_const();
+
         // Each component of the timestamp is in UTC time. Readers then apply the offset minutes
         // to derive the localized time.
-        let utc = self.date_time;
+        let utc = timestamp.date_time;
 
         // Ion encodes offsets in minutes while DateTime stores it in seconds.
-        let offset_seconds = self.offset.map(|o| o.utc_minus_local()).unwrap_or(0);
+        let offset_seconds = timestamp.offset.map(|o| o.utc_minus_local()).unwrap_or(0);
         let offset_minutes = (offset_seconds as f32 / SECONDS_PER_MINUTE).round() as i64;
-        VarInt::write_i64(&mut sink, offset_minutes)?;
-        VarUInt::write_u64(&mut sink, utc.year() as u64)?;
+        VarInt::write_i64(self, offset_minutes)?;
+        VarUInt::write_u64(self, utc.year() as u64)?;
```

Comment:
Not related to this change, but I think we should use a similar API pattern for these "encoders".  This is a reminder for me to cut an issue on that.

Reply:
Cut #277 to track this.

Reply:
Thanks. I left a note in it too re: their return value.